### PR TITLE
Catch wrong number of ranks to MPI unit tests.

### PR DESCRIPTION
**Pull request by PDoakORNL**
I also introduce a very simple abstraction Utilities/MPIExceptionWrapper.hpp to wrap unit test code using MPI exceptions thrown therein aren't lost and APP_ABORT is called.  This will help make sure we aren't chasing ghosts because someone has called the test directly with the "wrong" number of ranks or somehow overridden the rank count set by ADD_MPI_UNIT_TEST.

### CODE REVIEW

File: src/QMCDrivers/tests/test_WalkerControlMPI.cpp [hunk 1]
```diff
@@ -192,6 +201,10 @@ TEST_CASE("MPI WalkerControl population swap walkers", "[drivers][walker_control
     std::vector<int> count_after{1, 2, 2};
     test.testPopulationDiff(count_before, count_after);
   }
+                   };
```
Comment (ye-luo) [inline src/QMCDrivers/tests/test_WalkerControlMPI.cpp]: Needs formatting.

File: src/Utilities/tests/CMakeLists.txt [hunk 2]
```diff
@@ -23,3 +23,16 @@ TARGET_LINK_LIBRARIES(${UTEST_EXE} qmcutil ${QMC_UTIL_LIBS} ${MPI_LIBRARY})
...
 #ADD_TEST(NAME ${UTEST_NAME} COMMAND "${QMCPACK_UNIT_TEST_DIR}/${UTEST_EXE}")
 ADD_UNIT_TEST(${UTEST_NAME} "${QMCPACK_UNIT_TEST_DIR}/${UTEST_EXE}")
+
+IF(HAVE_MPI)
+  SET(UTEST_EXE test_${SRC_DIR}_mpi)
+  SET(UTEST_NAME deterministic-unit_test_${SRC_DIR}_mpi)
+  #this is dependent on the directory creation and sym linking of earlier driver tests
+  SET(MPI_UTILITY_TEST_SRC test_mpi_exception_wrapper.cpp)
+  ADD_EXECUTABLE(${UTEST_EXE} ${MPI_UTILITY_TEST_SRC})
+  USE_FAKE_RNG(${UTEST_EXE})
```
Comment (ye-luo) [inline src/Utilities/tests/CMakeLists.txt]: I don't think we need this RNG stuff here.

Comment (PDoakORNL) [reply src/Utilities/tests/CMakeLists.txt]: Will delete that

File: src/Utilities/MPIExceptionWrapper.hpp [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+#include "Message/Communicate.h"
+
+namespace qmcplusplus
+{
+
+class MPIExceptionWrapper
+{
+public:
+  MPIExceptionWrapper() {};
```
Comment (ye-luo) [inline src/Utilities/MPIExceptionWrapper.hpp]: In this case, I think there is no need of explicit empty constructor.

Comment (ye-luo) [reply src/Utilities/MPIExceptionWrapper.hpp]: Please also delete the std::cout in `~MCPopulation` and probably just delete the whole empty destructor.

**Comment by prckent** [comment]
@ye-luo Are you OK with these changes?

**Pull request merged**